crates/websocket/src/lib.rs

```diff
@@ -0,0 +1,380 @@
+        Ok(data)
+    }
+}
+
+/// Generic plaintext data codec that transmits UTF-8 strings as-is using
+/// WebSocket text messages.
+#[derive(Debug, Default)]
+pub struct Plaintext;
+
+impl DataCodec for Plaintext {
+    type MessageCodec = TextMessage;
+    type Payload = String;
+
+    fn encode(&self, data: Self::Payload) -> Result<Bytes, Error> {
+        Ok(data.into())
+    }
+
+    fn decode(&self, data: Bytes) -> Result<Self::Payload, Error> {
+        String::from_utf8(data.into()).map_err(Error::deserialization)
+    }
+}
+
+mod sealed {
+    use super::*;
+
+    /// Codec for encoding and decoding WebSocket messages.
+    ///
+    /// The only two useful implementations for consumers are [`BinaryMessage`]
+    /// and [`TextMessage`].
+    pub trait MessageCodec {
+        fn encode(data: Bytes) -> Result<Message, Error>;
+        fn decode(data: Message) -> Result<Bytes, Error>;
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error("Message is not binary")]
+struct InvalidBinaryError;
+
+/// WebSocket message codec for binary data. Encodes the payload into
+/// [`Message::Binary`].
+pub struct BinaryMessage;
+
+impl sealed::MessageCodec for BinaryMessage {
+    fn encode(data: Bytes) -> Result<Message, Error> {
+        Ok(Message::Binary(data))
+    }
+
+    fn decode(data: Message) -> Result<Bytes, Error> {
+        if let Message::Binary(data) = data {
+            Ok(data)
+        } else {
+            Err(Error::invalid_payload(InvalidBinaryError))
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error("Message is not UTF-8")]
+struct InvalidUtf8Error;
+
+/// WebSocket message codec for text data. Encodes the payload into
+/// [`Message::Text`].
+pub struct TextMessage;
+
+impl sealed::MessageCodec for TextMessage {
+    fn encode(data: Bytes) -> Result<Message, Error> {
+        String::from_utf8(data.into())
+            .map(Message::Text)
+            .map_err(|_| Error::invalid_payload(InvalidUtf8Error))
+    }
+
+    fn decode(data: Message) -> Result<Bytes, Error> {
+        if let Message::Text(data) = data {
+            Ok(data.into_bytes().into())
+        } else {
+            Err(Error::invalid_payload(InvalidBinaryError))
+        }
+    }
+}
+
+/// Builder for configuring and constructing a [`WebSocket`] instance.
+pub struct Builder<A, C, O> {
+    adapter: A,
+    codec: C,
+    observer: O,
+    config: Config,
+}
+
+impl Builder<(), (), ()> {
+    /// Create a new [`WebSocket`] builder instance.
+    pub fn new() -> Self {
+        Self {
+            adapter: (),
+            codec: (),
+            observer: (),
+            config: Default::default(),
+        }
+    }
+}
+
+impl Default for Builder<(), (), ()> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<A, C, O> Builder<A, C, O> {
+    /// Set the [`Adapter`] for the WebSocket.
+    pub fn adapter<T>(self, adapter: T) -> Builder<T, C, O>
+    where
+        T: Adapter,
```

Comment:
`websocket::Adapter` doesn't seem to represent what it actually is, as it's implemented for different external websocket implementations.
`Transport` maybe? or `Backend`?

Reply:
I don't like `Transport`, as it's too generic, and would prefer to leave it for a trait e.g. `trait Transport: Sink<...> + Stream<...> {}`. As for `Backend` vs `Adapter` I'm not really sure if one is better than the other. But I don't mind renaming it to `Backend`.

Any other options to consider?